complete/src/main/java/com/example/android/wearable/complicationsdatasource/CustomComplicationDataSourceService.kt

```diff
@@ -0,0 +1,155 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.complicationsdatasource
+
+import android.content.ComponentName
+import android.util.Log
+import androidx.wear.watchface.complications.data.ComplicationData
+import androidx.wear.watchface.complications.data.ComplicationType
+import androidx.wear.watchface.complications.data.LongTextComplicationData
+import androidx.wear.watchface.complications.data.PlainComplicationText
+import androidx.wear.watchface.complications.data.RangedValueComplicationData
+import androidx.wear.watchface.complications.data.ShortTextComplicationData
+import androidx.wear.watchface.complications.datasource.ComplicationRequest
+import androidx.wear.watchface.complications.datasource.SuspendingComplicationDataSourceService
+import com.example.android.wearable.complicationsdatasource.data.TAP_COUNTER_PREF_KEY
+import com.example.android.wearable.complicationsdatasource.data.dataStore
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import java.util.Locale
+
+/**
+ * Example watch face complication data source provides a number that can be incremented on tap.
+ *
+ * Note: This class uses the suspending variation of complication data source service to support
+ * async calls to the data layer, that is, to the DataStore saving the persistent values.
+ */
+class CustomComplicationDataSourceService : SuspendingComplicationDataSourceService() {
+
+    /*
+     * Called when a complication has been activated. The method is for any one-time
+     * (per complication) set-up.
+     *
+     * You can continue sending data for the active complicationId until onComplicationDeactivated()
+     * is called.
+     */
+    override fun onComplicationActivated(
+        complicationInstanceId: Int,
+        type: ComplicationType
+    ) {
+        Log.d(TAG, "onComplicationActivated(): $complicationInstanceId")
+    }
+
+    /*
+     * A request for representative preview data for the complication, for use in the editor UI.
+     * Preview data is assumed to be static per type. E.g. for a complication that displays the
+     * date and time of an event, rather than returning the real time it should return a fixed date
+     * and time such as 10:10 Aug 1st.
+     *
+     * This will be called on a background thread.
+     */
+    override fun getPreviewData(type: ComplicationType): ComplicationData {
+        return ShortTextComplicationData.Builder(
+        text = PlainComplicationText.Builder(text = "6!").build(),
+        contentDescription = PlainComplicationText
```

Comment:
Nit: this formatting ended up being a little weird

Reply:
Fixed all the formatting.